packages/scheduler/lib/scheduler.ts

```diff
@@ -163,12 +163,13 @@ export class Scheduler {
     /**
      * Fail a task
      * @param taskId - Task ID
+     * @param error - Json object representing the error
      * @returns Task
      * @example
-     * const failed = await scheduler.fail({ taskId: '00000000-0000-0000-0000-000000000000' });
+     * const failed = await scheduler.fail({ taskId: '00000000-0000-0000-0000-000000000000', error: {message: 'error'});
      */
-    public async fail({ taskId }: { taskId: string }): Promise<Result<Task>> {
-        const failed = await tasks.transitionState({ taskId, newState: 'FAILED' });
+    public async fail({ taskId, error }: { taskId: string; error: JsonValue }): Promise<Result<Task>> {
```

Comment:
Not directly related to this PR but I feel if you destructured `task` it would be more readable in the fail function

```
    public async fail({ taskId, error }: { taskId: string; error: JsonValue }): Promise<Result<Task>> {
        const failed = await tasks.transitionState({ taskId, newState: 'FAILED', output: error });
        if (failed.isOk()) {
            const task = failed.value;
            const { state, retryMax, retryCount, name, payload, groupKey, createdToStartedTimeoutSecs, startedToCompletedTimeoutSecs, heartbeatTimeoutSecs } =
                task;
            this.onCallbacks[state](task);
            // Create a new task if the task is retryable
            if (retryMax > retryCount) {
                const schedulingProps: SchedulingProps = {
                    scheduling: 'immediate',
                    taskProps: {
                        name,
                        payload,
                        groupKey,
                        retryMax,
                        retryCount: retryCount + 1,
                        createdToStartedTimeoutSecs,
                        startedToCompletedTimeoutSecs,
                        heartbeatTimeoutSecs
                    }
                };
                const res = await this.schedule(schedulingProps);
                if (res.isErr()) {
                    logger.error(`Error retrying task '${taskId}': ${stringifyError(res.error)}`);
                }
            }
        }
        return failed;
    }
```